bluesky/utils.py

```diff
@@ -1298,3 +1271,150 @@ def my_name(obj):
+    else:
+        import matplotlib
+        backend = matplotlib.get_backend().lower()
+        # if with a Qt backend, do the scary thing
+        if 'qt' in backend:
+            from matplotlib.backends.qt_compat import QtCore, QtWidgets
+            app = QtWidgets.QApplication.instance()
+            if app is None:
+                _qapp = app = QtWidgets.QApplication([b'bluesky'])
+            assert app is not None
+            event_loop = QtCore.QEventLoop()
+
+            def start_killer_thread():
+                def exit_loop():
+                    blocking_event.wait()
+                    # If the above wait ends quickly, we need to avoid the race
+                    # condition where this thread might try to exit the qApp
+                    # before it even starts.  Therefore, we use QTimer, below,
+                    # which will not start running until the qApp event loop is
+                    # running.
+                    event_loop.exit()
+
+                threading.Thread(target=exit_loop).start()
+
+            # https://www.riverbankcomputing.com/pipermail/pyqt/2015-March/035674.html
+            # adapted from code at
+            # https://bitbucket.org/tortoisehg/thg/commits/550e1df5fbad
+            if os.name == 'posix' and hasattr(signal, 'set_wakeup_fd'):
+                # Wake up Python interpreter via pipe so that SIGINT
+                # can be handled immediately.
+                # (http://qt-project.org/doc/qt-4.8/unix-signals.html)
+                # Updated docs:
+                # https://doc.qt.io/qt-5/unix-signals.html
+                import fcntl
+                rfd, wfd = os.pipe()
+                for fd in (rfd, wfd):
+                    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+                    fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)
+                wakeupsn = QtCore.QSocketNotifier(rfd,
+                                                  QtCore.QSocketNotifier.Read)
+                origwakeupfd = signal.set_wakeup_fd(wfd)
+
+                def cleanup():
+                    wakeupsn.setEnabled(False)
+                    rfd = wakeupsn.socket()
+                    wfd = signal.set_wakeup_fd(origwakeupfd)
+                    os.close(rfd)
+                    os.close(wfd)
+
+                def handleWakeup(inp):
+                    # here Python signal handler will be invoked
+                    # this book-keeping is to drain the pipe
+                    wakeupsn.setEnabled(False)
+                    rfd = wakeupsn.socket()
+                    try:
+                        os.read(rfd, 4096)
+                    except OSError as inst:
+                        print('failed to read wakeup fd: %s\n' % inst)
+
+                    wakeupsn.setEnabled(True)
+
+                wakeupsn.activated.connect(handleWakeup)
+
+            else:
+                # On Windows, non-blocking anonymous pipe or socket is
```

Comment:
Has any interactive testing been done on Windows with this?

Reply:
this was adapted from tortioushg which does get heavy windows use so I am optimistic.